### PR TITLE
Client Authentication - Client Secret Basic

### PR DESCRIPTION
**Pull request by anjuthomas**

### CODE REVIEW

File: README/clientSecretBasic.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+## Client Secret Basic
```
Comment (TMSCH) [inline README/clientSecretBasic.md]: I missed this on the previous one. Why do you create a readme file for each class? It doesn't contain any information specific to the class.

File: src/clientAuth/clientSecretBasic.js [hunk 2]
```diff
@@ -1,3 +1,9 @@
+/**
+ * @fileoverview One of the six different client authentication / authorization
```
Comment (TMSCH) [inline src/clientAuth/clientSecretBasic.js:2]: Be more specific. What does this client auth method does?

File: src/clientAuth/clientSecretBasic.js [hunk 3]
```diff
@@ -6,69 +12,65 @@ class ClientSecretBasic extends ClientAuthnMethod {
...
-      user = kwargs['user'];
+      user = kwargs.user;
     } else {
       user = cliInfo.client_id;
     }
-    if (Object.keys(httpArgs).indexOf('headers') === -1) {
-      httpArgs['headers'] = {};
-    }
-    let credentials = {};
+    httpArgs.headers = httpArgs.headers || {};
```
Comment (TMSCH) [inline src/clientAuth/clientSecretBasic.js]: Move this 2 lines below above where you set the Authorization header.

File: src/clientAuth/clientSecretBasic.js [hunk 4]
```diff
@@ -6,69 +12,65 @@ class ClientSecretBasic extends ClientAuthnMethod {
...
+      passwd = kwargs.password;
     } else {
-      if (httpArgs['password']) {
-        passwd = httpArgs['password'];
+      if (httpArgs.password) {
+        passwd = httpArgs.password;
       } else {
-        if (cis['client_secret']) {
-          passwd = cis['client_secret'];
+        if (cis.client_secret) {
```
Comment (TMSCH) [inline src/clientAuth/clientSecretBasic.js]: If `cis` is `null` this will raise error.

File: src/clientAuth/clientSecretBasic.js [hunk 5]
```diff
@@ -6,69 +12,65 @@ class ClientSecretBasic extends ClientAuthnMethod {
...
-    }
-    let credentials = {};
+    httpArgs.headers = httpArgs.headers || {};
+    const credentials = {};
     credentials[user] = passwd;
-    httpArgs['headers']['Authorization'] = credentials;
+    httpArgs.headers['Authorization'] = credentials;
     try {
-      delete cis['client_secret'];
+      delete cis.client_secret;
```
Comment (TMSCH) [inline src/clientAuth/clientSecretBasic.js]: Do not use try catch for such things. Check whether `cis` is not null.

File: src/clientAuth/clientSecretBasic.js [hunk 6]
```diff
@@ -6,69 +12,65 @@ class ClientSecretBasic extends ClientAuthnMethod {
...
+    if (cis.grant_type === 'authorization_code') {
+      if (!cis.client_id) {
+        if (cliInfo.client_id) {
+          cis.client_id = cliInfo.client_id;
+        } else {
           return;
         }
       }
     } else {
       let req = null;
```
Comment (TMSCH) [inline src/clientAuth/clientSecretBasic.js]: `const req = cis.client_id ? cis.client_id || false`

File: src/clientAuth/clientSecretBasic.js [hunk 7]
```diff
@@ -6,69 +12,65 @@ class ClientSecretBasic extends ClientAuthnMethod {
...
-      if (cis['client_id']) {
-        req = cis['client_id'];
+      if (cis.client_id) {
+        req = cis.client_id;
       } else {
         req = false;
       }
 
       if (!req) {
         try {
```
Comment (TMSCH) [inline src/clientAuth/clientSecretBasic.js]: Same here. Do not use try/catch.

File: src/clientAuth/clientSecretBasic.js [hunk 8]
```diff
@@ -6,69 +12,65 @@ class ClientSecretBasic extends ClientAuthnMethod {
...
    */
-  construct(cis, cliInfo, requestArgs, httpArgs, kwargs) {
-    cliInfo = cliInfo || null;
-    requestArgs = requestArgs || null;
-    httpArgs = httpArgs || null;
-
-    if (httpArgs == null) {
-      httpArgs = {};
-    }
+  construct(cis, cliInfo = null, requestArgs = null, httpArgs = {}, kwargs) {
```
Comment (TMSCH) [inline src/clientAuth/clientSecretBasic.js]: We'll have to discuss the method signature at some point. The logic to extract the relevant properties (password, client_secret, user, etc.) seems overly complex. I'd like to see some examples of how such API is used to understand why it's this way. Could you add a global readme with such examples? Thanks!

File: src/clientAuth/clientSecretBasic.js [hunk 9]
```diff
@@ -6,69 +12,65 @@ class ClientSecretBasic extends ClientAuthnMethod {
...
+          passwd = cis.client_secret;
         } else {
           passwd = cliInfo.client_secret;
         }
       }
     }
-
+    
     let user = null;
     if (kwargs) {
```
Comment (TMSCH) [inline src/clientAuth/clientSecretBasic.js]: `const user = kwargs ? kwargs.user || cliInfo.client_id;`

I'd even suggest:

`const user = kwargs && kwargs.user ? kwargs.user || cliInfo.client_id;`

Use the same method in other places.

**Review by TMSCH — CHANGES_REQUESTED** [review]

File: README/clientSecretBasic.md (hunk 1 above)
Comment (anjuthomas) [reply README/clientSecretBasic.md]: The auth0 team wanted me to add a readme for each pull request stating the intended usage.

File: src/clientAuth/clientSecretBasic.js (hunk 8 above)
Comment (anjuthomas) [reply src/clientAuth/clientSecretBasic.js]: The logic is based off of the python library so I would suggest discussing with Roland as well. Will add a README with examples to my TODO list. For now please refer to http://oiccli.readthedocs.io/en/latest/intro.html .

**Review by TMSCH — APPROVED** [review]

**Pull request merged**